### ISSUE TITLE
Backup and revert

### ISSUE THREAD

**Issue by JeroenVinke** [monterey-framework/cli#16]
We should backup the aurelia.json file and allow the user to restore the previous configuration (perhaps `au import --revert` ?)

**Comment by martonsagi** [comment]
We could create a checkpoint mechanism, which would allow users to move back and forth in configuration history.
Idea: revert command could also create a new checkpoint before reverting to selected state.

**List checkpoint**
`au import --revert list`

**Revert to previous state:**
`au import --revert`

**Revert to specific checkpoint:**
`au import --revert 12`

**Remove existing checkpoints to cleanup `backup` folder**
`au import --cleanup`

Checkpoint range: e.g. 1-999 (?). Filenames contain zeros only to keep accurate sorting by name.
Imagined directory structure (root: `aurelia_project/`):
* ./backup/
  * aurelia.json.001.bak
  * aurelia.json.002.bak
  * aurelia.json.003.bak
  * aurelia.json.004.bak
  * aurelia.json.005.bak
  * aurelia.json.006.bak
  * aurelia.json.007.bak
  * aurelia.json.008.bak

**Comment by JeroenVinke** [comment]
The use case for reverting is that you used `au import <package>` and found that it did not work as expected, and you want to revert this change, right? Do you think that a user will want to go back further? And how would you know what change has happened between backup 4 and 5? Also, manual changes of the aurelia.json file are not backupped, so that may create some confusion.

**Comment by vegarringdal** [comment]
With backup and revert...

We only need to revert if install fails, like normal npm install.
And we only need to have option to uninstall.

Being able to go back in history would be cool, but having install and uninstall to begin with is the most important ?
For history we would need some kinda git solution I think to make it really work

**Comment by adriatic** [comment]
@JeroenVinke and @vegarringdal make good points against a solution that on the surface looks very cool but it quickle becomes unmanageable. Even having the ability to list incremental differences will not help as these are notoriously difficult to parse for a human.

The actual syntax: `au import --revert` is not perfect as a "normal person" reading this command might see it as "null operator" (first import something - and then undo that), a more carefully observer would see that the absence of the package name means to undo the last import made..

It is important to understand effect of the `au import <package name>` command. Obviosly, the package importer tool being discussed here will effect the configuration files **and it will also change the content of `node_modules` folder** in a "transactional way". 

What is the actual mechanism executed when I issue `au import --revert` after `au import <package name>`  finished.

**Comment by JeroenVinke** [comment]
@adriatic 

> The actual syntax: au import --revert is not perfect as a "normal person" reading this command might see it as "null operator" (first import something - and then undo that), a more carefully observer would see that the absence of the package name means to undo the last import made..

That could be an issue yes. When `au import <package>` or `au install <package>`  finishes, we can explain that `au import --revert` can be used to revert the changes made by the importer. This should remove any confusion.

> Obviosly, the package importer tool being discussed here will effect the configuration files and it will also change the content of node_modules folder in a "transactional way".

Right now only the aurelia.json file will be modified by the importer. Before we make any changes we will backup the aurelia.json file. Then, `au import --revert` would simply replace the aurelia.json file with the one we backupped, effectively undoing the `au import` operation.

**Comment by adriatic** [comment]
> Right now only the aurelia.json file will be modified by the importer

Yes, that was the plan we discussed initially - however, it does not make sense to me now: one cannot know whether the modified `aurelia.json` file is correct or it has to be reverted until you run the application defined by this `aurelia.json` file. At that time, "reverting" the last import implies the need to delete the `node_modules` folder - the most expensive operation in the application building process there is.

I am not that dumb to believe how this issue escaped you, but stating that the `revert` operation takes less than a second is not OK - and it should encourage is to consider some innovation like replacing the `node-modules` with a real database. Now, this of course is a task for the company that builds npm - but I would not preclude the possibility that they might listen to us.

**Comment by JeroenVinke** [comment]
> At that time, "reverting" the last import implies the need to delete the node_modules folder - the most expensive operation in the application building process there is.

This would be true if we were talking about changes to the `package.json` file, not the `aurelia.json` file. `aurelia.json` simply tells aurelia-cli what packages to include in the bundle (and how to load the package). Changing this file does not result in a different `node_modules` directory, and therefore it does not have to be deleted and recreated.

**Comment by vegarringdal** [comment]
Sorry, havent had much time lately, and that time have gone into improving the grid.
Small question just to know whats what.

```au import <package>``` adds the needed code to aurelia.json from a package installed with ```npm install <package>, if it fails, it just revert to the aurelia.json before the command was launched ?

```au install <package>``` runs npm install and adds the needed code to aurelia.json ?

```au uninstall <package>``` does what ? just remove code from aurelia.json

**Comment by adriatic** [comment]
@JeroenVinke 
> `aurelia.json` simply tells `aurelia-cli` what packages to include in the bundle (and how to load the package).

Full recollection restored, thank you. This specific situation (not touching `node-modules` folder) is indeed the strong reason to keep `au import <package>` command decoupled from npm (see original Rob's spec)

**Comment by JeroenVinke** [comment]
@vegarringdal `au uninstall` does not exist yet, but will probably use the same importer strategies to remove a package as was used to install the package (more about the strategies can be found in [this document](https://aurelia-ui-toolkits.gitbooks.io/monterey-developer-manual/content/infrastructure/aurelia_import/design.html)). This means that the package itself has a chance to remove itself before we remove the configuration from aurelia.json.

`au install` will install the package (through npm/yarn/whatever) and will then call `au import`. `au import` simply takes the already installed package, and will change aurelia.json with the configuration for the package. `au import` will not install the package
